File: frontend/src/api/kpi.ts
/**
 * API 模块 — KPI 目标管理 (Phase 9)
 *
 * 对接后端 /api/v1/admin/kpi:
 *  - GET  /              列出全部目标
 *  - POST /              创建/更新目标 (后端走 upsert, 200)
 *  - GET  /achievement   达成率快照 (T-906 才用)
 *
 * RBAC: 后端 require_role(admin, manager); 前端入口同步收敛。
 */
import request from '@/api/request'

export type KpiScope = 'global' | 'department' | 'role'
export type KpiMetric = 'submit_rate' | 'avg_score' | 'blocker_resolve_days' | 'objective_completion'
export type KpiPeriod = 'weekly' | 'monthly' | 'quarterly'

export interface KpiTargetOut {
  id: number
  scope: KpiScope
  scope_value: string | null
  metric: KpiMetric
  target_value: number
  period: KpiPeriod
  created_at: string
  updated_at: string
  created_by: string | null
  tenant_id: string
}

export interface KpiTargetIn {
  scope: KpiScope
  scope_value: string | null
  metric: KpiMetric
  target_value: number
  period: KpiPeriod
}

export const listKpiTargets = (): Promise<KpiTargetOut[]> =>
  request.get('/admin/kpi/') as unknown as Promise<KpiTargetOut[]>

export const upsertKpiTarget = (payload: KpiTargetIn): Promise<KpiTargetOut> =>
  request.post('/admin/kpi/', payload) as unknown as Promise<KpiTargetOut>

export type AchievementStatus = 'on_track' | 'below_target' | 'no_data'

export interface KpiAchievementRow {
  scope: KpiScope
  scope_value: string | null
  metric: KpiMetric
  period: KpiPeriod
  target_value: number
  actual_value: number | null
  gap: number | null
  achievement_rate: number | null
  status: AchievementStatus
}

export interface KpiAchievementResponse {
  period: KpiPeriod
  snapshot_at: string
  rows: KpiAchievementRow[]
}

export const getKpiAchievement = (period?: KpiPeriod): Promise<KpiAchievementResponse> =>
  request.get('/admin/kpi/achievement', { params: period ? { period } : {} }) as unknown as Promise<KpiAchievementResponse>

File: backend/app/schemas/kpi.py
"""
app/schemas/kpi.py — Phase 9 KPI 目标与达成率 Pydantic V2 Schemas

参考已落地的 ORM Model（app/models/kpi_target.py）字段约束：
  - scope ∈ {global, department, job_title}
  - scope_value 可空（global 模式必空，其他模式必填）
  - metric ∈ {submit_rate, avg_score, blocker_resolve_days, objective_completion}
  - period ∈ {weekly, monthly, quarterly}
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Self

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.kpi_target import KpiMetric, KpiPeriod, KpiScope


class KpiTargetIn(BaseModel):
    scope: KpiScope
    scope_value: str | None = Field(None, max_length=50)
    metric: KpiMetric
    target_value: float = Field(..., gt=0)
    period: KpiPeriod = KpiPeriod.monthly

    @model_validator(mode="after")
    def validate_scope_value(self) -> Self:
        if self.scope == KpiScope.global_:
            if self.scope_value is not None:
                raise ValueError("scope=global 时 scope_value 必须为 None")
            return self

        if self.scope_value is None or not self.scope_value.strip():
            raise ValueError("非 global scope 时 scope_value 必须非空")

        self.scope_value = self.scope_value.strip()
        return self


class KpiTargetOut(BaseModel):
    id: int
    scope: KpiScope
    scope_value: str | None
    metric: KpiMetric
    target_value: float
    period: KpiPeriod
    created_at: datetime | None
    updated_at: datetime | None
    created_by: uuid.UUID | None
    tenant_id: str

    model_config = ConfigDict(from_attributes=True)


class KpiAchievementRow(BaseModel):
    scope: KpiScope
    scope_value: str | None
    metric: KpiMetric
    period: KpiPeriod
    target_value: float
    actual_value: float | None
    gap: float | None
    achievement_rate: float | None
    status: Literal["on_track", "below_target", "no_data"]


class KpiAchievementResponse(BaseModel):
    period: KpiPeriod
    snapshot_at: datetime
    rows: list[KpiAchievementRow]
